_posts/2016-09-29-Rust-1.12.md

```diff
@@ -0,0 +1,341 @@
+don't match:
+
+![A new mismatched trait
+error](/images/2016-09-Rust-1.12/mismatched-trait-error.png)
+
+Initially, this error design was built to aid borrow-checking errors, but we
+found, like the error above, the format can be broadly applied to a wide variety
+of errors. If you would like to learn more about the design, check out the
+[previous blog post on the subject][err].
+
+[err]: https://blog.rust-lang.org/2016/08/10/Shape-of-errors-to-come.html
+
+Finally, you can also get these errors as JSON with a flag. Remember that error
+we showed above, at the start of the post? Here's an example of attempting to
+compile that code while passing the `--error-format=json` flag:
+
+```bash
+$ rustc borrowck-assign-comp.rs --error-format=json
+{"message":"cannot assign to `p.x` because it is borrowed","level":"error","spans":[{"file_name":"borrowck-assign-comp.rs","byte_start":562,"byte_end":563,"line_start":15,"line_end":15,"column_start":14,"column_end":15,"is_primary":false,"text":[{"text":"    let q = &p;","highlight_start":14,"highlight_end":15}],"label":"borrow of `p.x` occurs here","suggested_replacement":null,"expansion":null}],"label":"assignment to borrowed `p.x` occurs here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
+{"message":"aborting due to previous error","code":null,"level":"error","spans":[],"children":[],"rendered":null}
+```
+
+We've actually elided a bit of this for brevity's sake, but you get the idea.
+This output is primarily for tooling; we are continuing to invest in supporting
+IDEs and other useful development tools. This output is a small part of that
+effort.
+
+#### MIR code generation
+
+The new Rust "mid-level IR", usually called "MIR", gives the compiler a simpler
+way to think about Rust code than it's previous way of operating entirely on the
+Rust abstract syntax tree. It makes analysis and optimizations possible that
+have historically been difficult to implement correctly. The first of many
+upcoming changes to the compiler enabled by MIR is a rewrite of the pass that
+generates LLVM IR, what `rustc` calls "translation", and after many months of
+effort the MIR-based backend has proved itself ready for prime-time.
+
+TODO: Improve code complexity: size of old trans vs size of new trans.
+
+While for the most part this new backend is simply laying the groundwork for
+future improvements, this release does include one long-awaited refinement to
+the run-time representation of Rust types that results in both compile-time wins
+and runtime wins: the removal of "drop flags", AKA "non-zeroing drop", as
+specified in [RFC 320].
+
+To maintain the semantics of Rust's `Drop` implementation, we always need to
+check whether a type must be dropped when it goes out of scope. Historically,
+this has been done by adding a special flag to all types that implement `Drop`,
+but this resulted in significant extra memory being used at runtime.
```

Comment:
The MIR backend is barely faster - however it's possible some of @nikomatsakis performance improvements to type/trait-related instrumentation to be in there, warrants further investigation.